typescript/sdk/src/providers/MultiProvider.ts

```diff
@@ -1,139 +1,350 @@
+  }
+
+  /**
+   * Get the ids for all chains known to this MultiProvider
+   */
+  getKnownDomainIds(): number[] {
+    return this.getKnownChainNames().map(this.getDomainId);
+  }
+
+  /**
+   * Get an Ethers provider for a given chain name, chain id, or domain id
+   */
+  tryGetProvider(chainNameOrId: ChainName | number): Provider | null {
+    const metadata = this.tryGetChainMetadata(chainNameOrId);
+    if (!metadata) return null;
+    const { name, chainId: id, publicRpcUrls } = metadata;
+
+    if (this.providers[name]) return this.providers[name];
+
+    if (TestChains.includes(name as CoreChainName)) {
+      this.providers[name] = new providers.JsonRpcProvider(
+        'http://localhost:8545',
+        31337,
+      );
+    } else if (publicRpcUrls.length && publicRpcUrls[0].http) {
+      this.providers[name] = new providers.JsonRpcProvider(
+        publicRpcUrls[0].http,
+        id,
+      );
+    } else {
+      return null;
     }
-    return chainConnection.provider;
+
+    return this.providers[name];
   }
 
   /**
-   * Get provider for a chain
-   * @returns value or null if chain value has not been set
+   * Get an Ethers provider for a given chain name, chain id, or domain id
+   * @throws if chain's metadata has not been set
    */
-  tryGetChainProvider(chain: Chain): ethers.providers.Provider | null {
-    return this.tryGet(chain)?.provider ?? null;
+  getProvider(chainNameOrId: ChainName | number): Provider {
+    const provider = this.tryGetProvider(chainNameOrId);
+    if (!provider)
+      throw new Error(`No chain metadata set for ${chainNameOrId}`);
+    return provider;
   }
 
   /**
-   * Get signer for a chain
-   * @throws if chain is invalid or has not been set
+   * Sets an Ethers provider for a given chain name, chain id, or domain id
+   * @throws if chain's metadata has not been set
    */
-  getChainSigner(chain: Chain): ethers.Signer {
-    const chainConnection = this.get(chain);
-    if (!chainConnection.signer) {
-      throw new Error(`No signer set for chain ${chain}`);
+  setProvider(chainNameOrId: ChainName | number, provider: Provider): Provider {
+    const chainName = this.getChainName(chainNameOrId);
+    this.providers[chainName] = provider;
+    return provider;
+  }
+
+  /**
+   * Sets Ethers providers for a set of chains
+   * @throws if chain's metadata has not been set
+   */
+  setProviders(providers: ChainMap<Provider>): void {
+    for (const chain of Object.keys(providers)) {
+      const chainName = this.getChainName(chain);
+      this.providers[chainName] = providers[chain];
     }
-    return chainConnection.signer;
   }
 
   /**
-   * Get signer for a chain
-   * @returns value or null if chain value has not been set
+   * Get an Ethers signer for a given chain name, chain id, or domain id
+   * If signer is not yet connected, it will be connected
+   */
+  tryGetSigner(chainNameOrId: ChainName | number): Signer | null {
```

Comment:
This is more of a question then a suggestion: When do we use `null` and when `undefined`?

Reply:
The subtleties of when to use null vs undefined in JS is more art than science. I would describe it as "Use null when something is *known* to be nothing and undefined when there may not be anything to know (e.g. a param that wasn't provided vs a param that was given and is explicitly null)